File: server/src/controllers/notificationController.ts
import { Request, Response } from 'express';
import { query } from '../config/db.js';
import { AuthRequest } from '../middleware/auth.js';
import { toMySQLDateTime } from '../utils/dataHelpers.js';

/**
 * Lấy danh sách notifications của user hiện tại
 */
export const getNotifications = async (req: Request, res: Response) => {
  try {
    const authReq = req as AuthRequest;
    const userId = authReq.userId;

    if (!userId) {
      return res.status(401).json({ error: 'Không có quyền truy cập' });
    }

    const { limit = 50, offset = 0, unreadOnly = false } = req.query;

    let sql = `
      SELECT 
        id, title, message, type, priority, is_read, metadata, created_at
      FROM notifications
      WHERE user_id = ?
    `;

    const params: any[] = [userId];

    if (unreadOnly === 'true') {
      sql += ' AND is_read = FALSE';
    }

    sql += ' ORDER BY created_at DESC LIMIT ? OFFSET ?';
    params.push(parseInt(limit as string), parseInt(offset as string));

    const notifications = await query<any[]>(sql, params);

    // Đếm tổng số unread
    const unreadCount = await query<any[]>(
      'SELECT COUNT(*) as count FROM notifications WHERE user_id = ? AND is_read = FALSE',
      [userId]
    );

    res.json({
      notifications: notifications.map((n: any) => ({
        id: n.id,
        title: n.title,
        message: n.message,
        type: n.type,
        priority: n.priority,
        read: n.is_read === 1 || n.is_read === true,
        metadata: n.metadata ? (typeof n.metadata === 'string' ? JSON.parse(n.metadata) : n.metadata) : null,
        createdAt: n.created_at,
      })),
      unreadCount: unreadCount[0]?.count || 0,
    });
  } catch (error: any) {
    console.error('Error fetching notifications:', error);
    res.status(500).json({ error: 'Không thể lấy danh sách thông báo' });
  }
};

/**
 * Đánh dấu notification là đã đọc
 */
export const markAsRead = async (req: Request, res: Response) => {
  try {
    const authReq = req as AuthRequest;
    const userId = authReq.userId;

    if (!userId) {
      return res.status(401).json({ error: 'Không có quyền truy cập' });
    }

    const { notificationId } = req.params;

    // Kiểm tra notification thuộc về user
    const notifications = await query<any[]>(
      'SELECT id FROM notifications WHERE id = ? AND user_id = ?',
      [notificationId, userId]
    );

    if (notifications.length === 0) {
      return res.status(404).json({ error: 'Không tìm thấy thông báo' });
    }

    await query(
      'UPDATE notifications SET is_read = TRUE WHERE id = ? AND user_id = ?',
      [notificationId, userId]
    );

    res.json({ success: true });
  } catch (error: any) {
    console.error('Error marking notification as read:', error);
    res.status(500).json({ error: 'Không thể đánh dấu đã đọc' });
  }
};

/**
 * Đánh dấu tất cả notifications là đã đọc
 */
export const markAllAsRead = async (req: Request, res: Response) => {
  try {
    const authReq = req as AuthRequest;
    const userId = authReq.userId;

    if (!userId) {
      return res.status(401).json({ error: 'Không có quyền truy cập' });
    }

    await query(
      'UPDATE notifications SET is_read = TRUE WHERE user_id = ? AND is_read = FALSE',
      [userId]
    );

    res.json({ success: true });
  } catch (error: any) {
    console.error('Error marking all notifications as read:', error);
    res.status(500).json({ error: 'Không thể đánh dấu tất cả đã đọc' });
  }
};

/**
 * Xóa notification
 */
export const deleteNotification = async (req: Request, res: Response) => {
  try {
    const authReq = req as AuthRequest;
    const userId = authReq.userId;

    if (!userId) {
      return res.status(401).json({ error: 'Không có quyền truy cập' });
    }

    const { notificationId } = req.params;

    // Kiểm tra notification thuộc về user
    const notifications = await query<any[]>(
      'SELECT id FROM notifications WHERE id = ? AND user_id = ?',
      [notificationId, userId]
    );

    if (notifications.length === 0) {
      return res.status(404).json({ error: 'Không tìm thấy thông báo' });
    }

    await query('DELETE FROM notifications WHERE id = ? AND user_id = ?', [
      notificationId,
      userId,
    ]);

    res.json({ success: true });
  } catch (error: any) {
    console.error('Error deleting notification:', error);
    res.status(500).json({ error: 'Không thể xóa thông báo' });
  }
};

File: server.net/Controllers/NotificationController.cs
using ConstructOs.Server.Config;
using ConstructOs.Server.Middleware;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Text.Json;

namespace ConstructOs.Server.Controllers;

[ApiController]
[Route("api/notifications")]
[Authorize]
public class NotificationController : ControllerBase
{
    private readonly IDatabaseService _db;

    public NotificationController(IDatabaseService db)
    {
        _db = db;
    }

    [HttpGet]
    public async Task<IActionResult> GetNotifications([FromQuery] int? limit, [FromQuery] int? offset, [FromQuery] bool? unreadOnly)
    {
        try
        {
            var userId = HttpContext.GetUserId();
            if (string.IsNullOrEmpty(userId))
            {
                return Unauthorized(new { error = "Không có quyền truy cập" });
            }

            var limitNum = limit ?? 50;
            var offsetNum = offset ?? 0;

            var sql = @"SELECT id, title, message, type, priority, is_read, metadata, created_at
                        FROM notifications
                        WHERE user_id = ?";
            var queryParams = new List<object> { userId };

            if (unreadOnly == true)
            {
                sql += " AND is_read = FALSE";
            }

            sql += " ORDER BY created_at DESC LIMIT ? OFFSET ?";
            queryParams.Add(limitNum);
            queryParams.Add(offsetNum);

            var notifications = await _db.QueryAsync<dynamic>(sql, queryParams.ToArray());

            var unreadCount = await _db.QueryFirstOrDefaultAsync<dynamic>(
                "SELECT COUNT(*) as count FROM notifications WHERE user_id = ? AND is_read = FALSE",
                new[] { userId }
            );

            var notificationsList = notifications.Select(n => new
            {
                id = n.id,
                title = n.title,
                message = n.message,
                type = n.type,
                priority = n.priority,
                read = Convert.ToBoolean(n.is_read ?? false),
                metadata = !string.IsNullOrEmpty(n.metadata?.ToString()) 
                    ? JsonSerializer.Deserialize<object>(n.metadata.ToString() ?? "{}") 
                    : null,
                createdAt = n.created_at
            }).ToList();

            return Ok(new
            {
                notifications = notificationsList,
                unreadCount = Convert.ToInt32(unreadCount?.count ?? 0)
            });
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error fetching notifications: {ex}");
            return StatusCode(500, new { error = "Không thể lấy danh sách thông báo" });
        }
    }

    [HttpPut("{notificationId}/read")]
    public async Task<IActionResult> MarkAsRead(string notificationId)
    {
        try
        {
            var userId = HttpContext.GetUserId();
            if (string.IsNullOrEmpty(userId))
            {
                return Unauthorized(new { error = "Không có quyền truy cập" });
            }

            var notifications = await _db.QueryAsync<dynamic>(
                "SELECT id FROM notifications WHERE id = ? AND user_id = ?",
                new[] { notificationId, userId }
            );

            if (!notifications.Any())
            {
                return NotFound(new { error = "Không tìm thấy thông báo" });
            }

            await _db.ExecuteAsync(
                "UPDATE notifications SET is_read = TRUE WHERE id = ? AND user_id = ?",
                new[] { notificationId, userId }
            );

            return Ok(new { success = true });
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error marking notification as read: {ex}");
            return StatusCode(500, new { error = "Không thể đánh dấu đã đọc" });
        }
    }

    [HttpPut("read-all")]
    public async Task<IActionResult> MarkAllAsRead()
    {
        try
        {
            var userId = HttpContext.GetUserId();
            if (string.IsNullOrEmpty(userId))
            {
                return Unauthorized(new { error = "Không có quyền truy cập" });
            }

            await _db.ExecuteAsync(
                "UPDATE notifications SET is_read = TRUE WHERE user_id = ? AND is_read = FALSE",
                new[] { userId }
            );

            return Ok(new { success = true });
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error marking all notifications as read: {ex}");
            return StatusCode(500, new { error = "Không thể đánh dấu tất cả đã đọc" });
        }
    }

    [HttpDelete("{notificationId}")]
    public async Task<IActionResult> DeleteNotification(string notificationId)
    {
        try
        {
            var userId = HttpContext.GetUserId();
            if (string.IsNullOrEmpty(userId))
            {
                return Unauthorized(new { error = "Không có quyền truy cập" });
            }

            var notifications = await _db.QueryAsync<dynamic>(
                "SELECT id FROM notifications WHERE id = ? AND user_id = ?",
                new[] { notificationId, userId }
            );

            if (!notifications.Any())
            {
                return NotFound(new { error = "Không tìm thấy thông báo" });
            }

            await _db.ExecuteAsync(
                "DELETE FROM notifications WHERE id = ? AND user_id = ?",
                new[] { notificationId, userId }
            );

            return Ok(new { success = true });
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error deleting notification: {ex}");
            return StatusCode(500, new { error = "Không thể xóa thông báo" });
        }
    }
}

